exes/mqtt-bridge/src/client.cpp

```diff
@@ -0,0 +1,218 @@
+#include "client.hpp"
+
+#include <cstdint>
+#include <functional>
+#include <memory>
+#include <optional>
+#include <string>
+#include <string_view>
+#include <variant>
+
+#include <sparkplug_b/sparkplug_b.pb.h>
+#include <async_mqtt/all.hpp>
+#include <boost/asio.hpp>
+#include <boost/asio/ip/tcp.hpp>
+
+#include "endpoint.hpp"
+#include "sparkplug_utils.hpp"
+
+namespace tfc::mqtt {
+
+client::client(asio::io_context& io_ctx, std::string_view mqtt_will_topic)
+    : io_ctx_(io_ctx), mqtt_will_topic_(mqtt_will_topic) {
+  using enum structs::ssl;
+  if (config_.value().ssl_active == yes) {
+    mqtt_client_ = std::make_unique<endpoint_client>(io_ctx_, yes);
+    asio::co_spawn(mqtt_client_->strand(), connect(), asio::detached);
+  } else if (config_.value().ssl_active == no) {
+    mqtt_client_ = std::make_unique<endpoint_client>(io_ctx_, no);
+    asio::co_spawn(mqtt_client_->strand(), connect(), asio::detached);
+  }
+}
+
+auto client::port_to_string(const std::variant<tfc::mqtt::config::port_e, uint16_t>& port) -> std::string {
+  return std::visit(
+      []<typename var>(var&& arg) { return std::to_string(static_cast<uint16_t>(std::forward<decltype(arg)>(arg))); }, port);
+}
+
+auto client::connect() -> asio::awaitable<void> {
+  logger_.trace("Resolving the MQTT broker address...");
+
+  asio::ip::tcp::socket resolve_sock{ io_ctx_ };
+  asio::ip::tcp::resolver res{ resolve_sock.get_executor() };
+
+  asio::ip::tcp::resolver::results_type resolved_ip =
+      co_await res.async_resolve(config_.value().address, port_to_string(config_.value().port), asio::use_awaitable);
+
+  while (true) {
+    auto connection_successful = co_await connect_and_handshake(resolved_ip);
+
+    if (connection_successful) {
+      co_return;
+    }
+
+    co_await mqtt_client_->close(asio::use_awaitable);
+    co_await asio::steady_timer{ io_ctx_, std::chrono::seconds{ 10 } }.async_wait(asio::use_awaitable);
+  }
+}
+
+auto client::connect_and_handshake(asio::ip::tcp::resolver::results_type const& resolved_ip) -> asio::awaitable<bool> {
+  logger_.trace("Resolved the MQTT broker address. Connecting...");
+
+  logger_.trace("Async connect for mqtt client");
+  co_await mqtt_client_->async_connect(resolved_ip);
+
+  logger_.trace("Setting SSL SNI");
+
+  mqtt_client_->set_sni_hostname(config_.value().address.c_str());
+
+  logger_.trace("Starting SSL handshake");
+
+  co_await mqtt_client_->async_handshake();
+
+  Payload payload;
+  payload.set_timestamp(sparkplug_utils::timestamp_milliseconds().count());
+
+  auto* metric = payload.add_metrics();
+
+  metric->set_name("bdSeq");
+  metric->set_timestamp(sparkplug_utils::timestamp_milliseconds().count());
+  metric->set_datatype(8);
+  metric->set_long_value(-1);
+
+  std::string payload_string;
+  payload.SerializeToString(&payload_string);
+  std::string_view payload_string_view{ payload_string };
+
+  payload.PrintDebugString();
```

Comment:
I think that the client getting the entire connect packet is too much but getting the will payload would be appropiate. If the caller would have to provide the entire connect_packet then it would have to have knowledge of variables such as the broker address, broker username and such. 

Moving the MQTT will payload to the caller, then client.cpp / client.hpp doesn't need any sparkplug dependancies.